### PR TITLE
Make StreamingDataset state_dict() more flexible

### PR DESCRIPTION
**Pull request by knighton**

### CODE REVIEW

File: streaming/base/dataset.py [hunk 1]
```diff
@@ -666,23 +666,29 @@ def __iter__(self) -> Iterator[Dict[str, Any]]:
...
-    def state_dict(self, sample_in_epoch: int) -> Dict[str, Any]:
+    def state_dict(self, num_samples: int, from_beginning: bool) -> Dict[str, Any]:
         """Get a dict containing training state (called from non-worker process).
 
         This is called on rank zero.
 
         Args:
-            sample_in_epoch (int): The number of samples processed so far in the current epoch.
+            num_samples (int): The number of samples processed so far in the current epoch.
+            from_beginning (int): Whether we are counting from start of this epoch, or start of
```
Comment (karan6181) [inline streaming/base/dataset.py:676]:
```suggestion
            from_beginning (int): Whether we are counting samples from start of this epoch, or start of
```

File: streaming/base/dataloader.py [hunk 2]
```diff
@@ -84,7 +86,7 @@ def load_state_dict(self, obj: Dict[str, Any]) -> None:
             obj (Dict[str, Any]): The state.
         """
         if isinstance(self.dataset, StreamingDataset):
-            return self.dataset.load_state_dict(obj)
+            self.dataset.load_state_dict(obj)
```
Comment (karan6181) [inline streaming/base/dataloader.py:89]: Why no `return` statement ?

File: streaming/base/dataloader.py [hunk 3]
```diff
@@ -72,7 +72,9 @@ def state_dict(self) -> Optional[Dict[str, Any]]:
         """
         if isinstance(self.dataset, StreamingDataset):
             world = World()
-            return self.dataset.state_dict(self.num_samples_yielded * world.num_ranks)
+            num_samples = self.num_samples_yielded * world.num_ranks
+            from_beginning = False
+            return self.dataset.state_dict(num_samples, from_beginning)
```
Comment (karan6181) [inline streaming/base/dataloader.py:77]: How about `return self.dataset.state_dict(num_samples, False)` ?

File: streaming/base/dataloader.py (hunk 2 above)
Comment (knighton) [reply streaming/base/dataloader.py:89]: it returns None; is superfluous

File: streaming/base/dataset.py [hunk 4]
```diff
@@ -666,23 +666,29 @@ def __iter__(self) -> Iterator[Dict[str, Any]]:
...
+                just this (potentially resumed) training run this epoch.
 
         Returns:
             Dict[str, Any]: The state.
         """
         world = World()
         epoch = self.next_epoch - 1
         epoch, offset = self._resume(world, epoch)
+        if from_beginning:
```
Comment (karan6181) [inline streaming/base/dataset.py:685]: maybe add a comment here saying that `from_beginning = True` is used when using `StreamingDataloader` ??

Comment (knighton) [reply streaming/base/dataset.py:688]: added note

File: streaming/base/dataloader.py (hunk 3 above)
Comment (knighton) [reply streaming/base/dataloader.py]: done

File: streaming/base/dataset.py (hunk 1 above)
Comment (knighton) [reply streaming/base/dataset.py]: done

**Review by karan6181 — APPROVED** [review]

**Pull request merged**